vm-setup/roles/v1aX_integration_test/tasks/remediation.yml

```diff
@@ -142,145 +197,279 @@
+
+  - name: Wait until "{{ NUMBER_OF_BMH }}" -1  machines become running.
+    k8s_info:
+      api_version: cluster.x-k8s.io/v1alpha3
+      kind: machines
+      namespace: "{{ NAMESPACE }}" 
+    register: machines
     retries: 200
     delay: 20
-    until: provisioned_machines.stdout == NUMBER_OF_BMH
+    until: 
+      - machines is succeeded 
+      - machines.resources | filter_phase("running") | length == (NUMBER_OF_BMH | int) - 1
 
   - name: Scale down the machinedeployment
-    shell: kubectl scale machinedeployment "{{ CLUSTER_NAME }}" -n "{{ NAMESPACE }}" --replicas=1
+    k8s:
+      api_version: cluster.x-k8s.io/{{ CAPI_VERSION }}
+      kind: MachineDeployment
+      name: "{{ CLUSTER_NAME }}"
+      namespace: "{{ NAMESPACE }}"
+      definition:
+        spec:
+          replicas: 1
 
   - name: Wait until one BMH is in ready state
-    shell: kubectl get bmh -n "{{ NAMESPACE }}" | grep -w ready | wc -l
+    k8s_info:
+       api_version: metal3.io/v1alpha1
+       kind: BareMetalHost
+       namespace: "{{ NAMESPACE }}"
+    register: bmhs
     retries: 200
     delay: 20
-    register: ready_hosts
-    until: ready_hosts.stdout == "1"
+    until:
+      - bmhs is succeeded 
+      - bmhs.resources | filter_provisioning("ready") | length == 1
 
     ## Start Metal3DataTemplate reference test
-  - name: Get the Metal3DataTemplate, Edit name , Add templateReference and apply it
-    shell: kubectl get m3dt -n "{{ NAMESPACE }}" "{{ CLUSTER_NAME }}"-workers-template  -ojson |
-              jq '.metadata.name="test-new-m3dt"' |
-              jq '.|del(.metadata|.managedFields,.uid,.resourceVersion,.ownerReferences,.finalizers)' |
-              jq '.|del(.status)' |
-              jq '.spec.templateReference="{{ CLUSTER_NAME }}-workers-template"' |
-              kubectl apply -f-
-
-  - name: Get the Metal3MachineTemplate, Edit name, Refer to new Metal3DataTemplate and apply it
-    shell: kubectl get m3mt -n "{{ NAMESPACE }}" "{{ CLUSTER_NAME }}"-workers -ojson |
-              jq '.spec.template.spec.dataTemplate.name="test-new-m3dt"' |
-              jq '.metadata.name="test-new-m3mt"' |
-              jq '.|del(.metadata|.managedFields,.uid,.resourceVersion,.ownerReferences,.finalizers)' |
-              jq '.|del(.status)' |
-              kubectl apply -f-
+  - name: Get the metal3datatemplate
+    k8s_info:
+     api_version: infrastructure.cluster.x-k8s.io/v1alpha4
+     kind: Metal3DataTemplate
+     name: "{{ CLUSTER_NAME }}-workers-template"
+     namespace: "{{ NAMESPACE }}"
+    register: m3dt
+
+  - name: Edit Metal3DataTemplate name and Add templateReference
+    k8s:
+     api_version: infrastructure.cluster.x-k8s.io/v1alpha4
+     kind: Metal3DataTemplate
+     namespace: "{{ NAMESPACE }}"
+     state: present
+     definition:
+       metadata:
+         name: "test-new-m3dt"
+       spec:
+         clusterName: "{{ CLUSTER_NAME }}"
+         templateReference: "{{ CLUSTER_NAME }}-workers-template"
```

Comment:
It is not quite obvious here what is going on. I think it would be better to really use all of the `m3dt` variable and only override the parts we want to change. As it is now one has to know that there are no other fields. Otherwise it looks like we are creating a "default" m3dt with some specific values from the old one and some values we set manually.

I'm not sure I'm explaining it in a good way. But what I mean is that I'd like to see something like this:
```yaml
k8s:
  api_version: infrastructure.cluster.x-k8s.io/v1alpha4
  kind: Metal3DataTemplate
  namespace: "{{ NAMESPACE }}"
  state: present
  definition: "{{ m3dt.resources[0] | replace_name_and_template_reference }}"
```
The point is not to have a filter that can do this, it is to define the new m3dt completely from the old one. It probably makes more sense to do it with `set_facts` than with a filter but I hope it shows what I mean.
The same goes for the m3mt below.

Reply:
Done using filters